caffe2/operators/numpy_tile_op.h

```diff
@@ -0,0 +1,116 @@
+#include "caffe2/core/operator.h"
+#include "caffe2/utils/math.h"
+
+namespace caffe2 {
+
+// Copy a Blob n times along a specified axis.
+template <class Context>
+class NumpyTileOp : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  NumpyTileOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws) {}
+  ~NumpyTileOp() {}
+
+  bool RunOnDevice() override {
+    const auto& input = Input(0);
+    const auto& repeats = Input(1);
+
+    // Check that the `repeats` tensor has the correct rank, has a number of
+    // elements equal to the number of axes of `input`.
+    CAFFE_ENFORCE(repeats.ndim() == 1, "repeats input must be a 1-d tensor");
+    CAFFE_ENFORCE(repeats.size() == input.ndim(), "repeats input have the same"
+                            " number of elements as `inputs` has dimensions.");
+    const int64_t *repeats_data = repeats.template data<int64_t>();
+    for (size_t i=0; i<repeats.size(); ++i) {
+      CAFFE_ENFORCE_GE(repeats_data[i], 0);
+    }
+
+    auto* output = Output(0);
+
+    // Alternate inputs and outputs between two buffers. Repeatedly apply the
+    // Tile kernel along each axis. Then copy out the resulting data into the
+    // output tensor.
+    Tensor<Context> *src = &buffer_a, *dst = &buffer_b;
+    src->CopyFrom(input);
+    vector<TIndex> output_dims(input.dims());
+    for (size_t i = 0; i < repeats.size(); ++i) {
+      if (repeats_data[i] == 1) {
+        continue;
+      }
+      // size up to (and not including) axis
+      const auto outer_dim = src->size_to_dim(i);
+      // size from axis up
+      const auto inner_dim = src->size_from_dim(i);
+
+      dst->Resize(outer_dim, inner_dim * repeats_data[i]);
+
+      /**
+       * How this works:
+       * Imagine a 2D tensor (matrix) of size 3x10, tiled 2 times.
+       * - Tiling along axis 0 (row) means copying the entire 3x10 Matrix 2
+       * times. outer_dim = 0, inner_dim = 30.
+       * - Tiling along axis 1 (column) means copying each row 2 times, then
+       * proceed to the next row, until the end. outer_dim = 3, inner_dim = 10.
+       */
+      const char* src_data = static_cast<const char*>(src->raw_data());
+      char* dst_data =
+          static_cast<char*>(dst->raw_mutable_data(src->meta()));
+
+      DoTile(
+          src->meta(),
+          src->itemsize(),
+          outer_dim,
+          inner_dim,
+          repeats_data[i],
+          src_data,
+          dst_data);
+
+        output_dims[i] *= repeats_data[i];
+        dst->Reshape(output_dims);
+
+        std::swap(src, dst);
+    }
+
+    // NB: because we have the swap at the end of the above loop, our real
+    // result tensor is going to live in *src when we reach this line
+    // whether we entered the loop or not :)
+    output->CopyFrom(*src);
+
+    // reshape output to be input tiled along the axis
+    output->Reshape(output_dims);
+
+    return true;
+  }
+
+ private:
+  void DoTile(
```

Comment:
probably proper implementation would special case for tiling the last dimension and use something like eigen to vectorize stuff. If it's not a perf bottleneck now - ship it